File: LeetCode/0863-All-Nodes-Distance-K-in-Binary-Tree.cpp
// 863. All Nodes Distance K in Binary Tree

class Solution {
public:
    void find_par(TreeNode* root, unordered_map<TreeNode*,TreeNode*>& par) {
        if(!root) return;
        if(root->left) {
            par[root->left] = root;
            find_par(root->left, par);
        }
        if(root->right) {
            par[root->right] = root;
            find_par(root->right, par);
        }
    }

    vector<int> distanceK(TreeNode* root, TreeNode* target, int k) {
        unordered_map<TreeNode*,TreeNode*> par;
        find_par(root, par);

        queue<TreeNode*> q;
        q.push(target);

        set<TreeNode*> vis;
        vis.insert(target);

        int dist = 0;
        vector<int> ans;
        while(!q.empty()) {
            int sz = q.size();
            if (dist == k) {
                while (sz--) {
                    ans.push_back(q.front()->val);
                    q.pop();
                }
                break;
            }

            while(sz--) {
                auto curr = q.front();
                q.pop();
                if(curr->left && !vis.count(curr->left)) {
                    q.push(curr->left);
                    vis.insert(curr->left);
                }
                if(curr->right && !vis.count(curr->right)) {
                    q.push(curr->right);
                    vis.insert(curr->right);
                }
                if(par.count(curr) && !vis.count(par[curr])) {
                    q.push(par[curr]);
                    vis.insert(par[curr]);
                }
            }

            dist++;
        }

        return ans;
    }
};

File: LeetCode/0863-All-Nodes-Distance-K-in-Binary-Tree.java
// 863. All Nodes Distance K in Binary Tree

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public Map<TreeNode, TreeNode> parent;

    public void getParent(TreeNode root) {
        if (root == null) return;
        if (root.left != null) {
            parent.put(root.left, root);
            getParent(root.left);
        }
        if (root.right != null) {
            parent.put(root.right, root);
            getParent(root.right);
        }
    }

    public List<Integer> distanceK(TreeNode root, TreeNode target, int k) {

        parent = new HashMap<>();
        getParent(root);

        List<Integer> ans = new ArrayList<>();
        Queue<TreeNode> q = new LinkedList<>();

        Set<TreeNode> vis = new HashSet<>();
        q.add(target);
        vis.add(target);
        int lvl = 0;

        while (!q.isEmpty()) {
            if (lvl == k) {
                while (!q.isEmpty()) {
                    ans.add(q.poll().val);
                }
            }
            int sz = q.size();
            while (sz-- > 0) {
                TreeNode curr = q.poll();
                if (curr.left != null && !vis.contains(curr.left)) {
                    vis.add(curr.left);
                    q.add(curr.left);
                }
                if (curr.right != null && !vis.contains(curr.right)) {
                    vis.add(curr.right);
                    q.add(curr.right);
                }
                if (parent.get(curr) != null && !vis.contains(parent.get(curr))) {
                    TreeNode par = parent.get(curr);
                    vis.add(par);
                    q.add(par);
                }
            }
            lvl++;
        }

        return ans;
    }
}
